src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs

```diff
@@ -159,6 +188,316 @@ private string GetRootPathWithGuid_NoLock()
             return new MetadataAsSourceFile(fileInfo.TemporaryFilePath, navigateLocation, documentName, documentTooltip);
         }
 
+        private async Task<Document> DecompileSymbolAsync(Document temporaryDocument, OptionSet options, ISymbol symbol, CancellationToken cancellationToken)
+        {
+            // Get the name of the type the symbol is in
+            var containingOrThis = symbol.GetContainingTypeOrThis();
+            var fullName = GetFullReflectionName(containingOrThis);
```

Comment:
I just tested if this works when trying to navigate to an inner class.
I discovered the following issues:
* For `A.B.ClassA.InnerClass` `GetFullReflectionName` returns `A.B.InnerClass`, which is completely wrong.
* The `FullTypeName(string)` ctor expects reflection syntax, so we'd have to use `A.B.ClassA+InnerClass`.
* `CSharpDecompiler.DecompileType` fails when it's passed the name of an inner class. I would say this is by design, but we could improve it by using the top-level type name stored in `FullTypeName`.

So there is two possible fixes:
a) Fix `GetFullReflectionName` to actually return the proper full name using reflection syntax and fix `CSharpDecompiler.DecompileType` to handle nested classes properly.
b) use something like:
```
            while (containingOrThis.ContainingType != null)
            {
                containingOrThis = containingOrThis.ContainingType;
            }
```
to unwrap the inner class before passing it to `GetFullReflectionName`.

I think I will fix the handling in `CSharpDecompiler.DecompileType` in any case.